### PR TITLE
Merge functionality for running external processes into core

### PR DESCRIPTION
**Pull request by NinjaWazzzabi**

### CODE REVIEW

File: core/src/main/src/ProcessRunner/ProcessRunner.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    /**
+     * Adds an observer that will receive the output from the process.
+     * Warning, all data that is outputted from the process will be sent,
+     * including all echoes that the process will output from using the
+     * {@link ProcessRunner#outputToScript(String)} method.
+     *
+     * @param observer Object that will receive the output.
+     */
+    public void addInputObserver(InputObserver observer);
+
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/ProcessRunner.java]: JavaDoc for removeInputObserver - Maybe you could rephrase "observer Object that will..." to something like "observer that will be unsubscribed"

File: core/src/main/src/ProcessRunner/ProcessRunner.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    public void start();
+
+    /**
+     * Send output dynamically to the process.
+     *
+     * @param s String that will be outputted to the process.
+     * @throws IOException Throws if process has yet to be started or if another output error has occurred.
+     */
+    public void outputToScript(String s) throws IOException;
+
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/ProcessRunner.java]: JavaDoc for forceCloseProcess:
Rephrasing the doc description to be more concise and clear. For example: "Once executed on the ProcessRunner-object, the ProcessRunner will be useless"

File: core/src/main/src/ProcessRunner/ProcessRunnerImpl.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package ProcessRunner;
+
+import java.io.*;
+import java.util.ArrayList;
+import java.util.List;
+
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/ProcessRunnerImpl.java:6]: I'd suggest that we rephrase the introductory JavaDoc to include the I/O definition

File: core/src/main/src/ProcessRunner/ProcessRunnerImpl.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+            System.out.println("-----------------------");
+        }
+
+        StreamReader reader = new StreamReader(p.getInputStream());
+        reader.setOnInputRead(this::alertObservers);
+        reader.start();
+        writer = new BufferedWriter(new OutputStreamWriter(p.getOutputStream()));
+    }
+
+    @Override
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/ProcessRunnerImpl.java:48]: Could the argument 's' be more descriptive in some way? Like "command" or "script", the argument is a bit obscure to me atm 🙂

File: core/src/main/src/ProcessRunner/ProcessRunnerImpl.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        inputObservers.add(observer);
+    }
+
+    @Override
+    public void removeInputObserver(InputObserver observer) {
+        inputObservers.remove(observer);
+    }
+
+    /**
+     * Alerts all of the current observers of the String that has been outputted from the process.
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/ProcessRunnerImpl.java]: Maybe rephrase the argument here as well to something more clear? Like "message" 

Also change "outputted" to "output", it's more preferred by the public

File: core/src/main/src/ProcessRunner/StreamReader.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.function.Consumer;
+
+/**
+ * Created by Virtuality.
+ */
+
+//Class is largely based on:
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/StreamReader.java]: Remove comment?

File: core/src/main/src/ProcessRunner/StreamReader.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            receivedValue = br.read();
+            if (receivedValue != 0) {
+                onInputRead.accept(String.valueOf(((char)receivedValue)));
+            }
+        }
+        System.out.println("StreamReader closed!");
+    }
+
+
+    //Bug found, will only return a string when a \n has been read.
```
Comment (felixnorden) [inline core/src/main/src/ProcessRunner/StreamReader.java]: Is it necessary when we have a char reader? If so, feel free to comment the purpose and UC of the method 😄

File: core/src/test/java/ProcessRunner/ProcessIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Created by Virtuality.
+ */
+class ProcessIntegrationTest {
+
+    @Test
+    void cmdReadTest() {
```
Comment (felixnorden) [inline core/src/test/java/ProcessRunner/ProcessIntegrationTest.java:19]: Maybe we should skip this before we integrate it into the system? We can't really be platform dependant on windows

File: core/src/test/java/ProcessRunner/ProcessIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        ProcessRunner process = ProcessFactory.createPythonProcess(pyScript.getAbsolutePath());
+
+        InputReceiver receiver = new InputReceiver();
+        process.addInputObserver(receiver);
+        process.start();
+
+        // 3. Wait for process to start
+        Thread.sleep(500);
+
+        // 3. Read the first word of the command output
```
Comment (felixnorden) [inline core/src/test/java/ProcessRunner/ProcessIntegrationTest.java]: Wrong numbering, should be 4.

File: core/src/test/java/ProcessRunner/ProcessIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        // First word should be "test"
+        assert sb.toString().equals("test");
+    }
+
+    @Test
+    void pythonWriteTest() throws IOException, InterruptedException {
+        // 1. Create temporary Python file
+        File pyScript = new File("pythonTemp.py");
```
Comment (felixnorden) [inline core/src/test/java/ProcessRunner/ProcessIntegrationTest.java:147]: Rename variable to bw or something that can be associated to the Buffered Writer class? bf seems qute unclear :)

File: core/src/test/java/ProcessRunner/ProcessIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+        process.addInputObserver(receiver);
+        process.start();
+
+        // 3. Wait for process to start
+        Thread.sleep(250);
+
+        // 4. Output "1" to the process
+        process.outputToScript("1");
+
+        // 5. Wait for process to handle the output
```
Comment (felixnorden) [inline core/src/test/java/ProcessRunner/ProcessIntegrationTest.java]: maybe clarify what is to be output? E.g refer to the script in the beginning or something alike 👍

File: core/src/test/java/ProcessRunner/ProcessIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+        }
+
+        // First word should be test
+        assert sb.toString().equals("test");
+    }
+
+
+    class InputReceiver implements InputObserver {
+
+        private final List<String> receivedInputs;
```
Comment (felixnorden) [inline core/src/test/java/ProcessRunner/ProcessIntegrationTest.java]: Rename to buffer

**Review by felixnorden — CHANGES_REQUESTED** [review]
Base fix: Please add a space between one line comment-descriptor and the comment. It will be a bit more readable. 😄 

Maybe adapt the method namings to adhere to verb-noun syntax?

**Review by NinjaWazzzabi — COMMENTED** [review]
Resolve all that was commented.

**Review by felixnorden — APPROVED** [review]
I Approve This Message 👍
